File: ruby/protein-translation/protein_translation.rb
module Translation
  MAP = {
    'AUG' => 'Methionine',
    'UUU' => 'Phenylalanine',
    'UUC' => 'Phenylalanine',
    'UUA' => 'Leucine',
    'UUG' => 'Leucine',
    'UCU' => 'Serine',
    'UCC' => 'Serine',
    'UCA' => 'Serine',
    'UCG' => 'Serine',
    'UAU' => 'Tyrosine',
    'UAC' => 'Tyrosine',
    'UGU' => 'Cysteine',
    'UGC' => 'Cysteine',
    'UGG' => 'Tryptophan',
    'UAA' => 'STOP',
    'UAG' => 'STOP',
    'UGA' => 'STOP'
  }.freeze

  def self.of_rna(strand)
    results = []
    results if strand.size < 3

    last = strand.size - 1
    start_at = 0
    end_at = 2

    while end_at <= strand.size

      codon = strand[start_at..end_at]
      protein = MAP.fetch(codon) { |key| raise InvalidCodonError, "Unknown codon: #{key}" }
      break if protein == 'STOP'

      results.push(MAP[codon])

      remaining = last - end_at
      raise InvalidCodonError, 'Incomplete sequence' if remaining.positive? && remaining < 3

      start_at = end_at + 1
      end_at += 3
    end

    results
  end
end

class InvalidCodonError < StandardError
end

File: lua/protein-translation/protein-translation.lua
local MAP <const> = {
	Methionine = { "AUG" },
	Phenylalanine = { "UUU", "UUC" },
	Leucine = { "UUA", "UUG" },
	Serine = { "UCU", "UCC", "UCA", "UCG" },
	Tyrosine = { "UAU", "UAC" },
	Cysteine = { "UGU", "UGC" },
	Tryptophan = { "UGG" },
	STOP = { "UAA", "UAG", "UGA" },
}

local function translate_codon(codon)
	for protein, codons in pairs(MAP) do
		for _i, value in ipairs(codons) do
			if value == codon then
				return protein
			end
		end
	end
	error("Unknown codon: " .. codon)
end

local function translate_rna_strand(rna_strand)
	if #rna_strand % 3 ~= 0 then
		error("Incomplete sequence!!")
	end

	local result = {}
	for codon in rna_strand:gmatch("...") do
		local protein = translate_codon(codon)
		if protein == "STOP" then
			break
		else
			table.insert(result, protein)
		end
	end

	return result
end

return { codon = translate_codon, rna_strand = translate_rna_strand }
